plugins/aggregators/histogram/histogram.go

```diff
@@ -0,0 +1,244 @@
+}
+
+// Add adds new hit to the buckets
+func (h *HistogramAggregator) Add(in telegraf.Metric) {
+	id := in.HashID()
+	agr, ok := h.cache[id]
+	if !ok {
+		agr = metricHistogramCollection{
+			metric:              in.Name(),
+			tags:                in.Tags(),
+			histogramCollection: make(map[string]counts),
+		}
+	}
+
+	for field, value := range in.Fields() {
+		buckets := h.getBuckets(in.Name(), field)
+		if buckets == nil {
+			continue
+		}
+
+		if agr.histogramCollection[field] == nil {
+			agr.histogramCollection[field] = make(counts, len(buckets)+1)
+		}
+
+		if value, ok := convert(value); ok {
+			index := sort.SearchFloat64s(buckets, value)
+			agr.histogramCollection[field][index]++
+		}
+	}
+
+	h.cache[id] = agr
+}
+
+// Push returns histogram values for metrics
+func (h *HistogramAggregator) Push(acc telegraf.Accumulator) {
+	for _, aggregate := range h.cache {
+		for field, counts := range aggregate.histogramCollection {
+
+			buckets := h.getBuckets(aggregate.metric, field)
+			count := int64(0)
+
+			for index, bucket := range buckets {
+				count += counts[index]
+				addFields(acc, aggregate, field, strconv.FormatFloat(bucket, 'f', 1, 64), count)
+			}
+
+			// the adding a value to the infinitive bucket
+			count += counts[len(counts)-1]
+			addFields(acc, aggregate, field, bucketInf, count)
+		}
+	}
+}
+
+// Reset does nothing, because we need to collect counts for a long time, otherwise if config parameter 'reset' has
+// small value, we will get a histogram with a small amount of the distribution.
+func (h *HistogramAggregator) Reset() {}
+
+// resetCache resets cached counts(hits) in the buckets
+func (h *HistogramAggregator) resetCache() {
+	h.cache = make(map[uint64]metricHistogramCollection)
+}
+
+// checkAndGetBuckets checks the order of buckets and returns them.
+func (h *HistogramAggregator) getBuckets(metric string, field string) []float64 {
+	if buckets, ok := h.buckets[metric][field]; ok {
+		return buckets
+	}
+
+	for _, config := range h.Configs {
+		if config.Metric == metric {
+			if !isBucketExists(field, config) {
+				continue
+			}
+
+			if _, ok := h.buckets[metric]; !ok {
+				h.buckets[metric] = make(bucketsByFields)
+			}
+
+			checkOrder(config.Buckets, metric, field)
+
+			h.buckets[metric][field] = config.Buckets
+		}
+	}
+
+	return h.buckets[metric][field]
+}
+
+// isBucketExists checks if buckets exists for the passed field
+func isBucketExists(field string, cfg config) bool {
+	if len(cfg.Fields) == 0 {
+		return true
+	}
+
+	for _, fl := range cfg.Fields {
+		if fl == field {
+			return true
+		}
+	}
+
+	return false
+}
+
+// addFields adds the field with specified tags to accumulator
+func addFields(acc telegraf.Accumulator, agr metricHistogramCollection, field string, bucketTagVal string, count int64) {
+	fields := map[string]interface{}{field + "_bucket": count}
+
+	tags := map[string]string{}
+	for key, val := range agr.tags {
+		tags[key] = val
+	}
+	tags[bucketTag] = bucketTagVal
```

Comment:
I cant remember what the problem was. But now it works correctly without any copying. I will remove this piece of code.